packages/parser/src/language/timeline/timeline.langium

```diff
@@ -0,0 +1,24 @@
+grammar Timeline
+import "../common/common";
+
+entry Timeline:
+  NEWLINE*
+  TIMELINE_KEYWORD NEWLINE*
+  TitleAndAccessibilities?
+  sections+=TimelineSection* | periods+=TimelinePeriod*
+;
+
+TimelineSection:
+  title=TIMELINE_SECTION_TITLE NEWLINE*
+  (periods+=TimelinePeriod)*
+;
+
+TimelinePeriod:
+  title=TIMELINE_PERIOD_TITLE NEWLINE* (":" events+=(TIMELINE_EVENT | TIMELINE_PERIOD_TITLE) NEWLINE*)*
+;
+
+terminal TIMELINE_KEYWORD: /timeline(?!\S)/;
+terminal TIMELINE_SECTION_TITLE: /timeline-section/;
+// TODO: figue out how to prevent `PERIOD_TITLE` to consume `EVENT`
```

Comment:
@msujew Is there a better way to handle this?

The context is that there are two terminal rule identical in pattern: `TIMELINE_PERIOD_TITLE` and `TIMELINE_EVENT`, but there is some differences of the return type of `ValueConverter` (replace `<br>` into `\n`).

`TIMELINE_PERIOD_TITLE` occurs before`TIMELINE_EVENT`, so when I want to parse `TIMELINE_EVENT` it would be tokenized as `TIMELINE_PERIOD_TITLE`, and `ValueConverter` won't do changes as expected.

Reply:
I'm facing the same issue with other diagrams, there's similar pattern in some rules and they're being tokenized as another rule instead of desired one:

```
journey
%% title: score: actors
title: 0: cat
```

the `cat` is tokenized similar to `title` although pattern is difference:

- title: `/[\n\r:]/+`
- actors: `/[\n\r,]/+`

Reply:
Uhm, I believe your token regexp are the same?

https://github.com/mermaid-js/mermaid/blob/822e4d30984cc6d4c1b4a51a1b2499e46719f20a/packages/parser/src/language/timeline/timelineMatcher.ts#L15

https://github.com/mermaid-js/mermaid/blob/822e4d30984cc6d4c1b4a51a1b2499e46719f20a/packages/parser/src/language/timeline/timelineMatcher.ts#L24

Reply:
I know, I already said that:

> The context is that there are two terminal rule identical in pattern:

The second message was referring to another new diagram I was trying to create it parser - `journey` diagram.

---

The current grammar is in this way, they have the same pattern:

```
timeline
event title: event period 1: event period 2
```

```mermaid
timeline
event title: event period 1: event period 2
```

Reply:
Ok, I'm not sure what the expected behavior is here. It works as intended. The tokenizer will always choose the first terminal that it encounters that fits a given input.

How it this solved in the bison grammar? I assume they use the same terminal for that?

Reply:
In current `jison` grammar, the `:` is part of rule itself:

<https://github.com/mermaid-js/mermaid/blob/38f2ff0bca4e0f816256d1fc11469e8502b16316/packages/mermaid/src/diagrams/timeline/parser/timeline.jison#L41>

Whereas in the current `langium` grammar isn't not:

<https://github.com/mermaid-js/mermaid/blob/822e4d30984cc6d4c1b4a51a1b2499e46719f20a/packages/parser/src/language/timeline/timeline.langium#L25>

Thanks for the highlight! Although not it won't make symbols as `operators` or `symbols` in generated `monarch`, but at least it's working as expected.